### ISSUE TITLE
can't report not-defined correctly

### ISSUE THREAD

**Issue by scutgreenlee** [eslint/eslint#14852]
**Tell us about your environment**

* **ESLint Version:7.31.0
* **Node Version:10.7.0
* **npm Version:6.1.0
* **Operating System:Tencent tlinux release 2.2

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**
default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
    "env": {
        "browser": true,
        "commonjs": true,
        "es2021": true
    },
    "extends": "eslint:recommended",
    "parserOptions": {
        "ecmaVersion": 12
    },
    "rules": {
    }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
"use strict"

test = "";
console.log(test);
let test;
```

```bash
eslint helloworld.js
```

**What did you expect to happen?**
It should report the error"'test' is not defined" because I use a not-defined variable before it is been declared.

**What actually happened? Please copy-paste the actual, raw output from ESLint.**
nothing output just like it is correct. Of course, the code cant't run correctly and throw an "ReferenceError: test is not defined"

**Steps to reproduce this issue:**

1. run npm init
2. create a js file helloworld.js. The code in file just like what I paste upper.
3. run eslint --init 
✔ How would you like to use ESLint? · problems
✔ What type of modules does your project use? · commonjs
✔ Which framework does your project use? · none
✔ Does your project use TypeScript? · No / Yes
✔ Where does your code run? · node
✔ What format do you want your config file to be in? · JSON
4.run eslint helloworld.js and nothing output
5.run node helloworld.js throw error

**Are you willing to submit a pull request to fix this bug?**
I am just a beginner so I guess I can't contribute anything, but I think eslint should remind me the incorrect variable using in the code I paste.

**Comment by ljharb** [comment]
There shouldn't be an error from `no-undef` - `let` and `const` are both hoisted.

Certainly at runtime, this will error, so I'd expect *some* rule to warn on it - but not no-undef.

**Comment by yeonjuan** [comment]
@scutgreenlee Hi :) It's not a bug as @ljharb described.
And there is a `no-use-before-define` rule to handle those cases.

- [online-demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdXNlLWJlZm9yZS1kZWZpbmU6IFwiZXJyb3JcIiovXG5cblwidXNlIHN0cmljdFwiXG5cbnRlc3QgPSBcIlwiO1xuY29uc29sZS5sb2codGVzdCk7XG5sZXQgdGVzdDtcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

```js
/*eslint no-use-before-define: "error"*/

"use strict"

test = ""; // Error 
console.log(test); // Error
let test;
```

**Comment by scutgreenlee** [comment]
> There shouldn't be an error from `no-undef` - `let` and `const` are both hoisted.
> 
> Certainly at runtime, this will error, so I'd expect _some_ rule to warn on it - but not no-undef.
Thank you for you explanation. It is certainly look like a **no-undef** error so I was confused.

**Comment by scutgreenlee** [comment]
> @scutgreenlee Hi :) It's not a bug as @ljharb described. And you can use [no-use-before-define](https://eslint.org/docs/rules/no-use-before-define) for those cases.
> 
> * [online-demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdXNlLWJlZm9yZS1kZWZpbmU6IFwiZXJyb3JcIiovXG5cblwidXNlIHN0cmljdFwiXG5cbnRlc3QgPSBcIlwiO1xuY29uc29sZS5sb2codGVzdCk7XG5sZXQgdGVzdDtcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)
> 
> ```js
> /*eslint no-use-before-define: "error"*/
> 
> "use strict"
> 
> test = ""; // Error 
> console.log(test); // Error
> let test;
> ```

Thank you for your solution very much. It does work.
